exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js

```diff
@@ -87,103 +86,90 @@ function createHeader(messageId) {
-        const body = CreateTransactionProtobuf.encode(data).finish();
-
-        body.forEach(element => {
-          buffer.push(element)
-        });
+        const body = CreateTransactionProtobuf.encode(data).finish()
+        const bufferWithBody = appendBuffer(new Uint8Array(buffer), body)
 
         // calculate payload and insert it into buffer
-        Exonum.Uint32.serialize(buffer.length + SIGNATURE_LENGTH, buffer, PAYLOD_SIZE_OFFSET)
+        Exonum.Uint32.serialize(bufferWithBody.length + SIGNATURE_LENGTH, bufferWithBody, PAYLOD_SIZE_OFFSET)
 
         // append signature
-        const signature = Exonum.sign(keyPair.secretKey, buffer)
+        const signature = Exonum.sign(keyPair.secretKey, bufferWithBody)
+        const signBuffer = appendBuffer(bufferWithBody, Exonum.hexadecimalToUint8Array(signature))
 
-        data.ownerPublicKey = keyPair.publicKey
-
-        return axios.post(TX_URL, {
-         protocol_version: PROTOCOL_VERSION,
-         service_id: SERVICE_ID,
-         message_id: TX_WALLET_ID,
-         signature: signature,
-         body: data
-        })
+        return axios.post(TX_URL, signBuffer)
       },
 
-      transfer(keyPair, receiver, amountToTransfer, seed) {
-        
+      transfer (keyPair, receiver, amountToTransfer, seed) {
         // serialize and append message header
         let buffer = createHeader(TX_TRANSFER_ID)
 
         const data = {
           seed: seed,
           senderId: Exonum.hexadecimalToUint8Array(keyPair.publicKey),
-          recipientId:  Exonum.hexadecimalToUint8Array(receiver),
+          recipientId: Exonum.hexadecimalToUint8Array(receiver),
           amount: amountToTransfer
         }
 
-        const body = TransferTransactionProtobuf.encode(data).finish();
-
-        body.forEach(element => {
-          buffer.push(element)
-        });
+        // serialize and append message body
+        const body = TransferTransactionProtobuf.encode(data).finish()
+        const bufferWithBody = appendBuffer(new Uint8Array(buffer), body)
 
         // calculate payload and insert it into buffer
-        Exonum.Uint32.serialize(buffer.length + SIGNATURE_LENGTH, buffer, PAYLOD_SIZE_OFFSET)
+        Exonum.Uint32.serialize(bufferWithBody.length + SIGNATURE_LENGTH, bufferWithBody, PAYLOD_SIZE_OFFSET)
 
         // append signature
-        const signature = Exonum.sign(keyPair.secretKey, buffer)
-
-        data.senderId = keyPair.publicKey
-        data.recipientId = receiver
-        return axios.post(TX_URL, {
-          protocol_version: PROTOCOL_VERSION,
-          service_id: SERVICE_ID,
-          message_id: TX_TRANSFER_ID,
-          signature: signature,
-          body: data
-        }).then(response => waitForAcceptance(keyPair.publicKey, response.data))
+        const signature = Exonum.sign(keyPair.secretKey, bufferWithBody)
+        const signBuffer = appendBuffer(bufferWithBody, Exonum.hexadecimalToUint8Array(signature))
```

Comment:
… or even
```
fun buildMessage(header: byte[], body: byte[], keyPair: KeyPair) {
  var unsignedMessage = header + body
  header[PAYLOAD_SIZE_OFFSET] = unsignedMessage.size()
  var signature = sign(unsignedMessage)
  ret unsignedMessage + signature
}
```

Reply:
+1 for naming like  unsignedMessage, message, signature